src/shims/unix/linux/sync.rs

```diff
@@ -29,9 +34,15 @@ pub fn futex<'tcx>(
     let op = this.read_scalar(op)?.to_i32()?;
     let val = this.read_scalar(val)?.to_i32()?;
 
+    // Storing this inside the allocation means that if an address gets reused by a new allocation,
+    // we'll use an independent futex queue for this... that seems acceptable.
```

Comment:
@rust-lang/miri this is a change in behavior. We used to index futexes entirely on their absolute address, so they stuck around even if the allocation backing that address disappears -- in fact one signal FUTEX_WAKE on unallocated memory. That works fine on Linux but is quite odd from a Rust AM perspective, and gives us no chance to ever clean up old futex state.

So this PR attaches the futex to the "virtual" address inside an allocation, meaning it disappears when the allocation is freed. Strictly speaking this means we will miss wakeups if the address gets reused, and then someone does FUTEX_WAKE and expects to wake up a thread that started blocking on the old address before it got deallocated... but that seems really contrived?

Reply:
I will factor this commit into a separate PR.